### PR TITLE
fix(security): no warning when sanitizing escaped html (#9392)

### PR DESCRIPTION
**Pull request by wkwiatek**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
https://github.com/angular/angular/issues/9392

**What is the new behavior?**
No warning when properly escaped html is passed to `sanitize`

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by mprobst** [comment]
Thanks for the PR!

This looks conceptually fine (we want to compare to the input, not to the intermediate parsed state), and the test looks good, but I don't understand why it works.

Shouldn't `safeHtml`, once we parsed and re-serialized it, also no longer contain the escaped entity, but rather the unicode code point directly? And if so, how or why is it different from the treatment of `unsafeHtml` (before your change), which was also just parsed and re-serialized?

Any idea what's going on there?

**Comment by wkwiatek** [comment]
The problem is produced by this line: https://github.com/angular/angular/pull/9413/files#diff-475057312e0ba8afc2f24a4df80d6eaaL249. It modifies the input so that it's no longer the initial value in some cases. `DOM.setInnerHTML(containerEl, unsafeHtml);` few lines above leaves the treatment to the browser so the value may be a little bit different than the one we passed in to the function.

**Comment by mprobst** [comment]
Assume this code, from a DevTools console session:

```ts
let d = document.createElement('div');
// <div>​</div>​
let unsafeHtml = 'hello &#x1f680;'; // the input to sanitize
d.innerHTML = unsafeHtml; // parse it
let safeHtml = d.innerHTML; // serialize it back
// "hello 🚀" -- safeHtml now contains the actual rocket character, unescaped.
safeHtml === unsafeHtml;
// false
```

So I don't understand why your PR works – I would expect `safeHtml` in your PR to contain the actual rocket character, not the Unicode escape, and I would expect `entryHtml` (maybe better called `unsafeInputHtml`) to no longer contain the escape.

So how does this wor?

**Comment by wkwiatek** [comment]
The code you've just pasted is actually fine and reflects the situation except of naming. Look what `safeHtml` is in the source: https://github.com/angular/angular/pull/9413/files#diff-475057312e0ba8afc2f24a4df80d6eaaL253. 

`safeHtml` at the end in condition is simply not related in any way with this whole stuff we're talking about. I refer to `do {} while  ()` block which I guess is only for mXSS protection. And was accidentally overwriting input of the function.

**Comment by mprobst** [comment]
Could you add this test case and see what happens?

```ts
    t.it('supports sanitizing escaped entities', () => {
      t.expect(sanitizeHtml('hellö')).toEqual('hellö');
      t.expect(logMsgs).toEqual([]);
    });
```

**Comment by wkwiatek** [comment]
No problem. 

Here's the output: 
```
Expected 'hell&#246;' to equal 'hellö'.
```
```
Expected [ 'WARNING: sanitizing HTML stripped some content.' ] to equal [  ].
```
Both I've expected before starting tests.

**Comment by mprobst** [comment]
So, do you think this change is worth it, given that we fundamentally cannot fix the problem?

**Comment by wkwiatek** [comment]
I think that fundamentally it works pretty well. Look, `sanitizeHtml('hellö')` gives you the output: `'hell&#246;'` which is fine (because really sanitizes the output) and also gives you warning that something was stripped. Then try to add a test like this: 

```ts
t.it('supports sanitizing escaped entities', () => {
  t.expect(sanitizeHtml('hell&#246;')).toEqual('hell&#246;');
  t.expect(logMsgs).toEqual([]);
});
```
Now input and output are **exactly** the same. Both tests in this PR will succeed. However in the master second will fail because of warning message that should not be logged in this case.  

I think it's still worth to add because in current version the information is just misleading.

Make sense?

**Comment by wkwiatek** [comment]
I think that fundamentally it works pretty well. Look, sanitizeHtml('hellö') gives you the output: 'hell&#246;' which is fine (because really sanitizes the output) and also gives you warning that something was stripped. Then try to add a test like this:

t.it('supports sanitizing escaped entities', () => {
  t.expect(sanitizeHtml('hell&#246;')).toEqual('hell&#246;');
  t.expect(logMsgs).toEqual([]);
});
Now input and output are exactly the same. Both tests in this PR will succeed. However in the master second will fail because of warning message that should not be logged in this case.

I think it's still worth to add because in current version the information is just misleading.

Make sense?

File: modules/@angular/platform-browser/src/security/html_sanitizer.ts [hunk 1]
```diff
@@ -223,11 +223,11 @@ function stripCustomNsAttrs(el: any) {
  * Sanitizes the given unsafe, untrusted HTML fragment, and returns HTML text that is safe to add to
  * the DOM in a browser environment.
  */
-export function sanitizeHtml(unsafeHtml: string): string {
+export function sanitizeHtml(entryHtml: string): string {
```
Comment (mprobst) [inline modules/@angular/platform-browser/src/security/html_sanitizer.ts]: nit: could you rename this to `unsafeHtmlInput`? It's kind of important here to keep track of what's safe and what isn't.

**Comment by mprobst** [comment]
We'll still warn people about effectively harmless changes (input changes that are not actually changing anything), but I can see your reasoning. Also, the change is harmless enough.

Could you fix the parameter name? Otherwise good to go.

File: modules/@angular/platform-browser/src/security/html_sanitizer.ts (hunk 1 above)
Comment (wkwiatek) [inline modules/@angular/platform-browser/src/security/html_sanitizer.ts]: Done

**Comment by mprobst** [comment]
Merged. Thanks for the contribution @wkwiatek!

**Comment by wkwiatek** [comment]
Thanks. You're welcome!

**Pull request merged**